### ISSUE TITLE
path_found is not getting freed, and also can be incorrectly resued

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#56]
- It is not getting freed when the dictionary is closed .
- Reusing this variable may cause a mess if a chdir() or dictionary_set_data_dir() is invoked between dictionary opens.

So I propose to free and nullify it in dictionary_delete() and dictionary_six().
I thought of 2 ways to do that:

1. Make it a non-static global, and use extern to access it in the needed 2 other places.
2. Make it a private static in a new function cache_data_path(), that will be used to get/free/set this variable.

Linas,  I can send a fix if you tell me which way to choose.

```
/* XXX static global variable used during dictionary open */
static char *path_found = NULL;

static void * dict_file_open(const char * fullname, void * user_data)
{
	const char * how = (const char *) user_data;
	FILE * fh =  fopen(fullname, how);
	if (fh && NULL == path_found)
	{
		path_found = strdup (fullname);
		prt_error("Info: Dictionary found at %s", fullname);
	}
	return (void *) fh;
}
```

**Comment by linas** [comment]
The more technically-correct fix is the second one ...

**Comment by ampli** [comment]
I just sent pull request #59.
While fixing it I noted an old problem: the dictpath is still searched after path_found is prepended to the file name.  It could cause inconsistencies of finding the rest of the files in a different data directory than the first one, and also if a particular file is missing in a data directory it could be found in another one.

Another slight problem is the addition of what seems to me a useless path component on "first_time_ever".
For example, on dictionary_create_lang("en") it adds an "en:" component at the start of the dictpath, so the rest of searches start with "en/en/..." when using this component.

I looked at the old revisions and it seems this two problems exist from the start of this code. On dictionary_create_lang("ru") it causes 512 fopen() failuers (and zero failures after fixing it). Surprisingly, saving these 500+ fopen()'s result only in a few percent saving of CPU time on dictionary open...

So I would like to send a fix that removes two code sections, as indicated below by #if 0:
1. Remove adding the apparently useless (unless I missed something)  prevpath component.
2. Invoke opencb() unconditionally. On a full path it will do the same. On a path prepended by path_found it will prevent the dictpath search. However, if a file is not found in a place it should be, but exists in another dictpath component directory, a mess will still happen. This cannot not be fixed without restructuring the code.

In case of a full path that is not found, the code currently tries to append it to the dictpath components,
a thing that is useless, and may theoretically cause looking in undesired locations. But I don't know if it worth fixing.

```
#if 0 /* This code looks to me as unneeded */
	/* Record the first path ever used, so that we can recycle it */
	if (first_time_ever)
	{
		strncpy (prevpath, filename, MAX_PATH_NAME);
		prevpath[MAX_PATH_NAME] = 0;
		pos = strrchr(prevpath, DIR_SEPARATOR);
		if (pos) *pos = 0;
		pos = strrchr(prevpath, DIR_SEPARATOR);
		if (pos) *(pos+1) = 0;
		first_time_ever = 0;
	}

	/* Look for absolute filename.
	 * Unix: starts with leading slash.
	 * Windows: starts with C:\  except that the drive letter may differ.
	 * Note that only native windows C library uses backslashs; mingw 
	 * seems to use forward-slash, from what I can tell.
	 */
	if ((filename[0] == '/')
#if defined(_WIN32) || defined(_MSC_VER) || defined(__MINGW32__)
		|| ((filename[1] == ':')
			 && ((filename[2] == '\\') || (filename[2] == '/')))
#endif
	)
#endif /* End of unneeded code */
	{
		/* fopen returns NULL if the file does not exist. */
		fp = opencb(filename, user_data);
		if (fp) return fp;
	}
```

**Comment by ampli** [comment]
Linas,
I can send you a complete fix for the dictionary open issue, that handles it in the intended way.
1. If the path is absolute, open the file. Return NULL if failed. (Don't continue to append the absolute path to the dictpath components like is currently done.)
2. Open all the files that should be in the same dictionary, always from the same directory (currently if an open of one of them fails, the dictpath is used, a thing that may find the missing file in DEFAULTPATH).
3. Still allow opening of "stand alone" files using object_open() (as done from the corpus stuff).
4. Invalidate path_found on each dictionary-create.

If such a patch is fine by you, please disregard and close pull request #59 and I will send you a replacement one that implements the above.

**Comment by linas** [comment]
I merged #59 before I read the above.   If you can implement your points 1.2.3.4. with *fewer* lines of code, i.e. in a *simpler* way than what is currently there, then, yes, I would be very interested in that.  If its larger and more complex, then no, its not really much of a fix. 

point 3. is quasi-important ... the corpus-ranking code is currently unused, and will become fully obsolete when the sqlite3 dictionary work is completed.  Or rather, the current corpus ranking code should be merged into the sqlite3 backend; they're storing more or less the same kind of information.

I guess I'm saying that point 3 can be abandoned & discarded.   

A very useful project would be to finish the sqlite3-based dictionary work.

I will get back to your word-graph code after I finish a deadline for a different project I'm working on... maybe a week or two more.

**Comment by ampli** [comment]
Hi Linas,
In #59 I forgot a debug printout. However, anyway I would like to send you shortly a dictionary open fix that will fix the issues above and also the following:

1. dictionary_get_data_dir() is invoked in each object_open(), and potentially does "heavy" things, especially on Windows. It is called 80+ times on en and 250+ times on ru. A change can be introduced to call it only once per dict open.

2. If binreloc is used, only the binreloc and the shared system directories are used, and the user doesn't have a straightforward way to use a private data directory.

Another thing that can be fixed, but not without adding code, is that  "lang" may be set to a meaningless value, eg. "**.**" (depending on the lang path given by the user). But this may not worth fixing so I would leave it as is.

Since the dictpath is no longer read from the environment, and I guess there are no plans to add such a feature again, I would remove the usage of PATH_SEPARATOR, and just use an array of paths. I would also make the path caching inside object_open() (removing the newly added  cache_data_path()).

All of this will make the code shorter and simpler.

**Comment by linas** [comment]
Sounds good to me.  I don't know who uses binreloc .. maybe that was used to create the debian packages, I'm not sure.  Or maybe Apple...

**Comment by ampli** [comment]
> A very useful project would be to finish the sqlite3-based dictionary work.

Can you please summarize (in another issue maybe) the goals of the sqlite3-based dictionary addition, the current status of the code, and what is needed to be finished?  I can then see if I can complete it.

BTW, while I looked at dictionary_create_from_db(), I found some small problems, that I would like to send a fix for them later:
1. dict->lang = lang; But lang can be a dynamic memory that is freed in dictionary_create_default_lang().
2. object_open() of the db can fail but this is not checked. So it returns a non-null dict pointer on failure.
In dictionary_create_lang() , this value is check, and if it is NULL (can happen if I make this fix), a text-based dictionary open is tried.
**The question is** what we would really do if dictionary_create_from_db(() fails, while a db file exists (since it is not tried if a db file doesn't exist) - just return NULL or try a text-based dictionary open?
3. I also noted that that the db file is checked for 0 length in 2 places:
- db_open()
- file_exists(), that is only called by check_db().
So in one place the check can be remove (and "rb" is in POSIX, a thing that can simplify additional code)..

**Comment by ampli** [comment]
Problem 1 (storing future reference to potentially freed memory) exists also in dictionary_six_str().
In addition, the code in utilities.c is very careful to use DIR_SEPARAOR  "\\" on windows.
However, the code in dictionary_six_str() and dictionary_create_from_db() checks for a hard-coded "/".
I would like to use there DIR_SEPARAOR instead, and to that end move its definition from utilities.c to utilities.h.

**Comment by linas** [comment]
-- created issue #62 for the sql design goals

-- regarding DIR_SEPARAOR -- it seems that windows works, today. So, I guess the hard-coed / somehow doesn't affect windows? Or its not used? Or some compatibility layer handles it ???

**Comment by ampli** [comment]
The argument if  dictionary_create_lang() can be a path to the desired dictionary, when its last component is the language name, which is actually the directory component in which the dictionary files are found. This last component is extracted and assigned to dict->lang. The dict->lang variable is currently used in the code for spelling (and previously used in other places) and is also to returned by dictionary_get_lang().

The first link-parser argument is directly transferred to dictionary_create_lang(), and thus can be a path (absolute or relative).

The language extraction of dict->lang is currently done by strrchr() of '/'. This works well when the argument is something like mydatadir/en, but would break on native Windows link-parser if mydatadir\en would be used.

I all the time (on Linux) use such private dictionaries for testing, but I guess nobody tried it on a natively compiled Windows application, or didn't mind that the language is then set to "mydatadir\\en". This can be easily fixed by using DIR_SEPARATOR.

I raise this issue because I am touching the dictionary-path code, and I thought it would be a good idea to fix all the problems I encounter (in a different commit each).

**Comment by ampli** [comment]
I forgot to add:
When I need to test several versions of dictionaries of the same language, I put them in directories like en1, en2 etc. But the current code then sets dict->lang incorrectly (to en1 etc.). Also if the last component of the dictionary directory is not related to the language code, all is lost (like when invoking link-parser from the dictionary directory and using "**.**" as argument).

This may hint that the language code should maybe read from the dictionary itself (maybe optionally - if exists there, else still set from the last directory component).

**Comment by linas** [comment]
I've sometimes thought about putting the language code in the dict, but I never found a strong argument about why it should be there.

I believe in design minimalism: the best code is the smallest, simplest, easiest-to-understand code.  I sometimes fail in  this, but that's my motto ...

**Comment by ampli** [comment]
Regarding the previous point, putting the language code in the dictionary also has the benefit of being able to mention the language "dialect", like "en_XX", for spell checking.  Currently this is hard-coded.

Another thing I would like to change is the search path for dictionaries. I think it currently doesn't make much sense in its try to support system, private and source directory dictionaries:

1. In case there is a system installation of LG, the DEFAULTPATH (a system location) is usually used, since the search path has DEFAULTPATH **before** the private locations "**.:data:..:../data:**".
This seems unreasonable - private location should be first, else they are not usable in a normal way (see below)
So the user even cannot use a lang path of mytdatadir/en" to specify a private dictionary under the local directory. **But** if a bogus absolute lang path of "/mydatadir/en" is used , then object_open() fails to find it (no such a directory at /), it then failes to find it when it appends it to the absolute path DEFAULTPATH, and then the relative path "." is used, so "mydatadir/en" relative to the current directory is actually used!

If binreloc is used, currently the only way to use a private dictionary is by using a lang like ../../../../../somthing/en ...
The current method also has a problem when trying to use a just compiled LG version (without installation) while another one is already installed (need to use a full path of the dict in the source directory, a thing that is not obvious).

2. In case there is **no** system installation of LG, the private locations are searched so there is no problem,unless there is a leftover system directory when removing pakages on systems which don't remove configuration files. This also will benefit from searching private locations first.

Hence I would like to change the dictpath to search the private paths first, and use the same basic path in binreloc and non--binreloc versions (of course the binreloc version will still include the binreloc component too in dictpath). I think that in normal use (of an installed LG) the result is compatible (the system path will still be used as intended). But the user has the added benefit of conveniently use private dicts and also the possibility to test a version from the installation directory even if there is an installed version.

**Comment by linas** [comment]
OK, this all sounds reasonable to me ...

**Comment by linas** [comment]
I mean, I'm not sure what else to say: sure, I don't see any particular problems ...

**Comment by ampli** [comment]
I have checked my modified dictopen" code, also on Windows (MSVC, Win8, cmd.exe console), and it seems fine.
It looks like paths with \ and / can be mixed with no problems, and even several \ and / in a rows work fine
(I don't know if this was so in previous Windows versions). However, such mixes, when happens, are not a result of my modifications.

There is some inconsistency that I would like to change in the code:
When word files are included with **/en/words/words.n.1**, object_open() gets the path without the leading **/**. However, this leading **/** stropping doesn't occur with #include directives (**#include "/ru/morph.dict"**).

The result is a not-nice code and a not-nice behavior, since this is the only reason object_open() has to retry a full-path open if it fails, and this causes it to try to prepend the dictpath also to real full paths that fail.

This is from the debug code - the first is a failed fopen():
```
link-grammar: Info: 1: object_open() trying /ru/suffix.dict=(nil)
link-grammar: Info: 2: object_open() trying ..//ru/suffix.dict=0x603600022c00
```

I tried to strip this **/**, and everything is still fine, while the code can be simplified and its behavior is more correct.  So I think this **/** should indeed be stripped too.
